### PR TITLE
BufferWhileBy: emit empty array when primary source finished first (#128)

### PR DESCRIPTION
**Pull request by shamansir**
See #128

### CODE REVIEW

**Comment by rpominov** [comment]
I actually thought of making this as an option this time, rather than introducing a breaking change. Should've done the same for `filterBy`, but haven't had that thought at the right time :(

Let's add option `emitEmpty` alongside with others (`flushOnEnd` & `flushOnChange `), `false` by default.
There will be an inconsistency with `filterBy`, that has that behavior by default, but we'll fix this in the future.

What do you think?

**Comment by shamansir** [comment]
Yeah, that's probably better than breaking everything. Though these two, `bufferBy`/`bufferWhileBy`, will be most customizable among others :).

**Comment by shamansir** [comment]
Done, I suppose, see new change :)

**Comment by rpominov** [comment]
Also could you add a test case with `{flushEmpty: true, flushOnChange: true}`, when the controlling stream "chnages" before any value from main stream so the result stream emits `[]`?

**Comment by rpominov** [comment]
The new test case is same as test with emitting `[]` on end case (which you've added earlier).

There is two cases now when the result stream can emit `[]`:
 - when the main stream ends,
 - when controlling stream changes from `true` to `false` (simply emits `false` actually).

We have a test case for the first one, but not for the second.

**Comment by shamansir** [comment]
```
it 'should flush empty buffer and then end when secondary ends (w/ {flushOnChange: true, emitEmpty: true})', ->
      expect(stream().bufferWhileBy(send(stream(), ['<end>']), {flushOnChange: true, emitEmpty: true})).toEmit [{current: []}, '<end:current>']
      a = stream()
      b = stream()
      expect(a.bufferWhileBy(b, {flushOnChange: true, emitEmpty: true})).toEmit [[], '<end>'], -> send(b, ['<end>'])
```

Is this one right? It fails in my case, should I fix it?

**Comment by shamansir** [comment]
(fails since emits just an empty array without `end` event)

**Comment by shamansir** [comment]
With a lot of "intersecting" options, it's becomes a bit harder for one's mind to get what actually should happen :)

**Comment by rpominov** [comment]
> Is this one right?

Not quite, its should be `send(b, [true, false])`

> fails since emits only empty array without end event

It's not necessarily have to end after emitting `[]` (see explanation below)

> With a lot of "intersecting" options, it's becomes a bit harder for one's mind to get what actually should happen :)

Yeah, sorry. I'll try to explain how it works. But perhaps you are right that we should simplify it somehow for easier understanding (maybe in v3.0 #103).

So here how it works. It buffers all the values from main stream, and may flush the buffer in 3 cases:

 - On a value from main stream, if latest value from controlling stream was `false`.
 - On a `false` value from controlling stream, if `{flushOnChange: true}`.
 - On end of main stream, unless `{flushOnEnd: false}`.

And normally in all those 3 cases we check if buffer not empty before emit. But if `{emitEmpty: true}`, we don't do that check. Also in the first case the buffer always has at least 1 value, because there just was a value from main stream, so `emitEmpty` does't matter here.

Also note that in the second case after emitting `[]`, the result stream doesn't end, why should it?

Hope that'll help :)

**Comment by shamansir** [comment]
Yes, that's a great explanation, thank you! I think I got this right this time.

**Comment by shamansir** [comment]
> On a value from main stream, if latest value from controlling stream was `false`
if `flushOnChange` is `false`, I suppose :)

**Comment by shamansir** [comment]
I'll be able to fix it in the evening, I hope there's no rush :blush:

**Comment by rpominov** [comment]
> if `flushOnChange` is false, I suppose :)

No, `flushOnChange` also doesn't matter here. On a value from main it puts the new value to the buffer and then immediately flushes the buffer.  


With `{flushOnChange: false}`:

```
foo:     ----1----2----3----4----5----6----7----8X
bar:     -----------f---------t---------fX

result:  --------------•----•--------------•----•X
               [1, 2, 3]  [4]      [5, 6, 7]  [8]
```

With `{flushOnChange: true}`:

```
foo:     ----1----2----3----4----5----6----7----8X
bar:     -----------f---------t---------fX

result:  -----------•--•----•-----------•--•----•X
               [1, 2] [3]  [4]     [5, 6] [7]  [8]
```

> I'll be able to fix it in the evening, I hope there's no rush :blush:

Of course, no problem :)

**Comment by shamansir** [comment]
Hope now I got things right. So I send `true` and `false` to a secondary stream (actually, a property), and since primary stream sent no events before, I check if it emitted empty array. I also wanted to clarify what does the first line in the test case and a difference with the last line. As I see it, first one checks `Kefir.bufferWhileBy` and last one checks `stream.bufferWhileBy(otherStream)`. Is what I see correct?)

File: test/specs/buffer-while-by.coffee [hunk 1]
```diff
@@ -20,12 +20,24 @@ describe 'bufferWhileBy', ->
...
       b = stream()
       expect(a.bufferWhileBy(b)).toEmit ['<end>'], -> send(a, ['<end>'])
 
+    it 'should flush empty buffer and then end when primary ends (w/ {emitEmpty: true})', ->
+      expect(send(stream(), ['<end>']).bufferWhileBy(stream(), {emitEmpty: true})).toEmit [{current: []}, '<end:current>']
+      a = stream()
+      b = stream()
+      expect(a.bufferWhileBy(b, {emitEmpty: true})).toEmit [[], '<end>'], -> send(a, ['<end>'])
+
+    it 'should flush empty buffer and then end when secondary ends (w/ {flushOnChange: true, emitEmpty: true})', ->
```
Comment (rpominov) [inline test/specs/buffer-while-by.coffee]: The test seems ok, but the title should be "should flush empty buffer when secondary emits false ...".

**Comment by rpominov** [comment]
> Also I wanted to clarify what does the first line in the test case and a difference with the last line. As I see it, first one checks Kefir.bufferWhileBy and last one checks stream.bufferWhileBy(otherStream). Is what I see correct?)

Not sure what you mean, we don't have `Kefir.bufferWhileBy`

It's all now looks good to me except for the test title.

**Comment by shamansir** [comment]
> Not sure what you mean, we don't have Kefir.bufferWhileBy

True, I somehow thought every stream instance-bound method has a "backup" as a static method. Though, yet I found `exports.stream`/`prop` in `test-helpers`, I feel there is a major difference between the first line and the last line in the test case, but I'd like to clarify, what is it? Or does first line tests if current value was properly emitted, and last one does not? I mean these two:

```coffeescript
expect(stream().bufferWhileBy(send(prop(), [true, false]), {flushOnChange: true, emitEmpty: true})).toEmit [{current: []}]
expect(a.bufferWhileBy(b, {flushOnChange: true, emitEmpty: true})).toEmit [[]], -> send(b, [true, false])
```

**Comment by rpominov** [comment]
> Or does first line tests if current value was properly emitted

Yeah, basically

In `expect(observable).toEmit(expectedLog, callback)` the callback is optional. It subscribes to `observable` and save events to a log; then if callback provided, calls the callback; then unsubscribe end compare log with `expectedLog`. So without callback it only captures current values.

It should be `send(prop(), [false])` in first line, btw. Both `send(prop(), [true, false])` and `send(prop(), [false])` create a property with current value `false`.

**Comment by shamansir** [comment]
:pray:

**Comment by rpominov** [comment]
Merged, thanks! 

I'll release a new version soon.

**Comment by rpominov** [comment]
Just published `2.8.0`. Sorry it took so long, busy week :bee:

**Comment by shamansir** [comment]
Thank you a lot! Will it soon be at `npm` or is it there already? No rush, just asking :)

**Comment by rpominov** [comment]
Yeah, it's on npm :)

**Comment by shamansir** [comment]
Great! Thanks again! :)

**Pull request merged**